### PR TITLE
Migrate to MV3 (Chrome, Firefox, Edge)

### LINKED ISSUE #565: MV3 Compatibility
<Do not use this to ask about lost codes or accounts>
<More info here: https://authenticator.cc/docs/en/lost-codes>
<You can also open an issue by Tweeting with #AuthenticatorFeedback>

https://blog.chromium.org/2020/12/manifest-v3-now-available-on-m88-beta.html

MV3 removes background scripts, which makes it impossible to store data in memory, but we (and other password managers) need this for key storage. We should look at how other password managers plan to handle this change.

https://crbug.com/930235

### PR DESCRIPTION
**Pull request by Sneezry**
Fixes #565

- [x] Move to `chrome.storage.session` to cache password
- [x] Move to `chrome.scripting` to inject scan QR script
- [x] Replace `content_security_policy` with `content_security_policy.extension_pages` or `content_security_policy.sandbox` as appropriate.
- [x] Replace `setTimeout` with `chrome.alarms` API for auto-lock
- [x] Replace `XMLHttpRequest` with `fetch` in backend agent
- [x] Migrate LocalStorage to `chrome.storage.local`
- [x] <del>Migrate `Image` and `canvas` to `OffscreenCanvas`(?) for service worker use</del> Move QR decode to content script
- [ ] E2E test
  + [ ] Chrome 
	  + [x] Can create new accounts
	  + [x] Can create new accounts by scanning QR codes
	  + [x] Can import existing accounts
	  + [x] Can migrate data with existing installed MV2 extension
	  + [x] Can set password
	  + [x] Can click to copy
	  + [x] Can click to auto fill
	  + [ ] Can auto backup to 3rd party storage
	  + [x] Auto-lock works as expected
	  + [x] Force lock works as expected
	  + [x] Context menu works as expected
	  + [x] Policy configuration works as expected
	  + [x] Storage area migration works as expected
	  + [ ] TBD
  + [ ] Firefox
	  + [x] Can create new accounts
	  + [x] Can create new accounts by scanning QR codes
	  + [x] Can import existing accounts
	  + [x] Can migrate data with existing installed MV2 extension
	  + [x] Can set password
	  + [x] Can click to copy
	  + [x] Can click to auto fill
	  + [x] Can auto backup to 3rd party storage
	  + [x] Auto-lock works as expected
	  + [x] Force lock works as expected
	  + [ ] ~Context menu works as expected~
	  + [ ] Policy configuration works as expected
	  + [x] Storage area migration works as expected
	  + [ ] TBD
  + [ ] Edge
	  + [x] Can create new accounts
	  + [x] Can create new accounts by scanning QR codes
	  + [x] Can import existing accounts
	  + [x] Can migrate data with existing installed MV2 extension
	  + [x] Can set password
	  + [x] Can click to copy
	  + [x] Can click to auto fill
	  + [ ] Can auto backup to 3rd party storage
	  + [x] Auto-lock works as expected
	  + [x] Force lock works as expected
	  + [x] Context menu works as expected
	  + [ ] Policy configuration works as expected
	  + [x] Storage area migration works as expected
	  + [ ] TBD

## Known issues

- ~Backup export failed when setting a password~
- ~Browser sync is disabled by default, need to double confirm if it is as expected~

### CODE REVIEW

**Comment by Sneezry** [comment]
Some scenario test is still on the way, but the code should be ready for review.

**Comment by Sneezry** [comment]
This PR only contains Chrome parts. Firefox and Edge MV3 migration should be in other PRs.

**Comment by Sneezry** [comment]
@rebornix Chrome, Firefox, and Edge are starting to migrate to MV3, MV2 will stop working in the next few months. There are many breaking changes in MV3, Safari may also have some updates. Once this PR has been merged, the Safari build may be broken.

**Comment by mymindstorm** [comment]
Just FYI, I might not be able to get to this until next Wednesday.

**Comment by mymindstorm** [comment]
I'm getting 

```
/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:133
                if(isError) throw e;
                            ^

Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:71:19)
    at Object.createHash (node:crypto:133:10)
    at module.exports (/home/mymindstorm/Projects/Authenticator-1/node_modules/webpack/lib/util/createHash.js:135:53)
    at NormalModule._initBuildHash (/home/mymindstorm/Projects/Authenticator-1/node_modules/webpack/lib/NormalModule.js:417:16)
    at handleParseError (/home/mymindstorm/Projects/Authenticator-1/node_modules/webpack/lib/NormalModule.js:471:10)
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/webpack/lib/NormalModule.js:503:5
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/webpack/lib/NormalModule.js:358:12
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:373:3
    at iterateNormalLoaders (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:214:10)
    at iterateNormalLoaders (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:221:10)
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:236:3
    at context.callback (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:111:13)
    at makeSourceMapAndFinish (/home/mymindstorm/Projects/Authenticator-1/node_modules/ts-loader/dist/index.js:90:5)
    at successLoader (/home/mymindstorm/Projects/Authenticator-1/node_modules/ts-loader/dist/index.js:68:9)
    at Object.loader (/home/mymindstorm/Projects/Authenticator-1/node_modules/ts-loader/dist/index.js:22:12)
    at LOADER_EXECUTION (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:119:14)
    at runSyncOrAsync (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:120:4)
    at iterateNormalLoaders (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:232:2)
    at Array.<anonymous> (/home/mymindstorm/Projects/Authenticator-1/node_modules/loader-runner/lib/LoaderRunner.js:205:4)
    at Storage.finished (/home/mymindstorm/Projects/Authenticator-1/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:55:16)
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:91:9
    at /home/mymindstorm/Projects/Authenticator-1/node_modules/graceful-fs/graceful-fs.js:123:16
    at FSReqCallback.readFileAfterClose [as oncomplete] (node:internal/fs/read_file_context:68:3) {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}

Node.js v18.12.1
```

when building. It goes away with `export NODE_OPTIONS=--openssl-legacy-provider`.

 I think this might be fixed if we use a newer version of loader-runner (seems to be a dependency of webpack). Would that be easy to do here?


---

The service worker isn't registering properly on my Chrome: `Service worker registration failed. Status code: 15`.

image

---

Keyboard shortcuts seem to not be working due to manifest issue:

image

**Comment by Sneezry** [comment]
Can you load this build locally?
[chrome.zip](https://github.com/Authenticator-Extension/Authenticator/files/10789323/chrome.zip)

**Comment by mymindstorm** [comment]
That build has same issues with the service worker and keyboard commands. (Chromium Version 110.0.5481.77)

**Comment by Sneezry** [comment]
Interesting... It works fine on my machine though...

image

image

image

**Comment by mymindstorm** [comment]
Interesting, I'll investigate further this weekend. What is your chrome version?


-------- Original message --------
From: Zhe Li ***@***.***>
Date: 2/20/23 9:47 PM (GMT-06:00)
To: Authenticator-Extension/Authenticator ***@***.***>
Cc: Brendan Early ***@***.***>, Review requested ***@***.***>
Subject: Re: [Authenticator-Extension/Authenticator] Migrate to Chrome MV3 (PR #1009)


Interesting... It works fine on my machine though...

[image]<https://user-images.githubusercontent.com/1621293/220242490-8e0dbd82-9afa-483d-b62b-1e09b59de843.png>

[image]<https://user-images.githubusercontent.com/1621293/220242531-f1ca9d29-bc67-4728-ac96-e99b04c89066.png>

[image]<https://user-images.githubusercontent.com/1621293/220242564-d99fe754-55a7-483d-97e9-3d7b66cf1b47.png>

**Comment by mymindstorm** [comment]
oops, I see that you've attached it.

**Comment by Sneezry** [comment]
I might find the root cause of the service work register issue. 

> Lots of API inside chrome are present only if your extension's manifest.json lists its associated permission name in "permissions" or in a special key
> https://stackoverflow.com/a/73789284

The contextMenu is in optional permission, which means `chrome.contextMenu` should be undefined before the permission granted.

We can move it to required permission since I see no security risk about that.

@mymindstorm do you have any concern?

**Comment by mymindstorm** [comment]
Sorry for the lateness on this. I will get to it as soon as I can.

**Comment by mymindstorm** [comment]
I agree, I don't see a security concern there.

**Comment by mymindstorm** [comment]
I think I've resolved most of the issues I found. Please make sure to test on your end too to make sure I didn't miss anything.

**Comment by mymindstorm** [comment]
I am leaving the tests broken, I ran out of time to fix them.

**Comment by Sneezry** [comment]
> I think I've resolved most of the issues I found. Please make sure to test on your end too to make sure I didn't miss anything.

Awesome! It works w/o problems on my side.

File: manifests/manifest-chrome.json [hunk 1]
```diff
@@ -41,32 +41,36 @@
...
+    ],
+    "optional_host_permissions": [
+      "https://www.google.com/",
+      "https://*.dropboxapi.com/*",
+      "https://www.googleapis.com/*",
+      "https://accounts.google.com/o/oauth2/revoke",
+      "https://graph.microsoft.com/me/*",
+      "https://login.microsoftonline.com/common/oauth2/v2.0/token"
     ],
     "offline_enabled": true,
```
Comment (DNin01) [inline manifests/manifest-chrome.json]: I can't find any information that suggests that `offline_enabled` is still supported by Manifest V3. Compare these pages:
https://developer.chrome.com/docs/extensions/mv2/manifest/
https://developer.chrome.com/docs/extensions/mv3/manifest/

[The page for this property on MDN Web Docs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/offline_enabled) also lists the supported manifest version as 2 only.

Comment (Sneezry) [reply manifests/manifest-chrome.json]: Thanks for the information.

**Comment by mymindstorm** [comment]
I spent a few hours working on tests. They technically work again. Except for the fact that I had to disable the UI ones until I have more time to work out the kinks of MV3 disallowing all unsafe-eval and some weird webpack dependency issues. At least it is in an overall better state than it was previously.

**Comment by Sneezry** [comment]
Firefox should now be operational. However, end-to-end testing has not yet been completed.

**Comment by WofWca** [comment]
Hey! So, the extension seems to be working alright on Manifest V3, it's only tests (i.e. non-production code) that are missing? Am I reading this correctly?

**Comment by Sneezry** [comment]
> Hey! So, the extension seems to be working alright on Manifest V3, it's only tests (i.e. non-production code) that are missing? Am I reading this correctly?

The end-to-end test is not completed. If all tests pass, we can deliver it to the public.

**Comment by mymindstorm** [comment]
Tests work again! We lost the ability for code coverage reporting, but that was half broken due to vue anyway.

image

File: src/models/storage.ts [hunk 2]
```diff
@@ -48,28 +59,33 @@ export class BrowserStorage {
...
       return new Promise((resolve) => {
-        resolve(localStorage.storageLocation);
+        resolve(LocalStorage.storageLocation);
         return;
       });
     }
   }
 
   // TODO: promise this
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (mymindstorm) [inline src/models/storage.ts]: remove these comments

File: src/models/storage.ts [hunk 3]
```diff
@@ -48,28 +59,33 @@ export class BrowserStorage {
...
-  static async get(callback: (items: { [key: string]: any }) => void) {
+  static async get() {
     const storageLocation = await this.getStorageLocation();
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (mymindstorm) [inline src/models/storage.ts]: remove this comment

**Review by mymindstorm — CHANGES_REQUESTED** [review]
I really don't like how `LocalStorage` works. 

 - Putting it as a global everywhere makes it easy to misuse (e.g. forget to load/store data from/to `chrome.storage` again). This should all be hidden behind a class. It's too easy to make mistakes with the current way.
 - Using the global var as a sort-of cache introduces the possibility of race conditions
 - We should name is something more descriptive, like `userSettings`
 - We know exactly what should be in `LocalStorage`/`userSettings`. There should be an interface for it instead of `any`.

For example, something like:

```ts
// UserSettings.d.ts
declare type StorageLocation = "sync" | "local";

interface UserSettings {
    storageLocation?: StorageLocation;
}

// LocalUserSettings.ts
export class LocalUserSettings {
  static async getUserSettings(): Promise<UserSettings> {
    return (await chrome.storage.local.get("userSettings"))?.userSettings || {};
  }

  static async setUserSettings(userSettings: UserSettings) {
    let mergedSettings = await getUserSettings();
    Object.assign(mergedSettings, userSettings);
    await chrome.storage.local.set({ userSettings: mergedSettings });
  }
}
```

What are your thoughts on this?

**Comment by mymindstorm** [comment]
Also, the migration for localStorage broke in Firefox. Will investigate when I have more time.

**Comment by Sneezry** [comment]
> I really don't like how `LocalStorage` works.
> 
> * Putting it as a global everywhere makes it easy to misuse (e.g. forget to load/store data from/to `chrome.storage` again). This should all be hidden behind a class. It's too easy to make mistakes with the current way.
> * Using the global var as a sort-of cache introduces the possibility of race conditions
> * We should name is something more descriptive, like `userSettings`
> * We know exactly what should be in `LocalStorage`/`userSettings`. There should be an interface for it instead of `any`.
> 
> For example, something like:
> 
> ```ts
> // UserSettings.d.ts
> declare type StorageLocation = "sync" | "local";
> 
> interface UserSettings {
>     storageLocation?: StorageLocation;
> }
> 
> // LocalUserSettings.ts
> export class LocalUserSettings {
>   static async getUserSettings(): Promise<UserSettings> {
>     return (await chrome.storage.local.get("userSettings"))?.userSettings || {};
>   }
> 
>   static async setUserSettings(userSettings: UserSettings) {
>     let mergedSettings = await getUserSettings();
>     Object.assign(mergedSettings, userSettings);
>     await chrome.storage.local.set({ userSettings: mergedSettings });
>   }
> }
> ```
> 
> What are your thoughts on this?

That makes sense. I will update the code later.

**Comment by mymindstorm** [comment]
Ok. I will be unavailable for the next week and a half, but can help out after that.

**Comment by Sneezry** [comment]
@mymindstorm I have updated the PR. Let's complete this PR before end of the next week (5/26).

**Comment by Sneezry** [comment]
> Also, the migration for localStorage broke in Firefox. Will investigate when I have more time.

I didn't meet any error when migrating from MV2. Could you test the latest code to verify if the issue has been resolved?

**Comment by mymindstorm** [comment]
Sure. I've been sick since last week so it might take me some time to get to.


-------- Original message --------
From: Zhe Li ***@***.***>
Date: 5/19/24 11:47 PM (GMT-06:00)
To: Authenticator-Extension/Authenticator ***@***.***>
Cc: Brendan Early ***@***.***>, Mention ***@***.***>
Subject: Re: [Authenticator-Extension/Authenticator] Migrate to MV3 (Chrome, Firefox, Edge) (PR #1009)


Also, the migration for localStorage broke in Firefox. Will investigate when I have more time.

I didn't meet any error when migrating from MV2. Could you test the latest code to verify if the issue has been resolved?

**Comment by Sneezry** [comment]
Oh, take care and recover soon!

Brendan Early ***@***.***>于2024年5月20日 周一20:46写道：

**Comment by Sneezry** [comment]
Considering Google is required all extensions to migrate to MV3 by June, and leaving a week for CWS review, I will complete this PR by 5/26 anyway.

**Comment by mymindstorm** [comment]
Just FYI, starting to work on this. Will probably push something within 3-6 hours.

**Review by mymindstorm — APPROVED** [review]
From what I've seen it appears to work fine, outside of that crash I found. Although, I want to start automating E2E tests at some point.

**Pull request merged**